[UNIT] Final review of Unit 2.1 on smolagents - release 25/2/25 (#195)
huggingface/agents-course: units/en/unit2/smolagents/why_use_smolagents.mdx

![smolagents banner](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/smolagents/license_to_call.png)
# Why use smolagents

> In this module, we will explore the pros and cons of using [smolagents](https://huggingface.co/blog/smolagents), helping you make an informed decision about whether it's the right framework for your needs.

Reviewer: ```suggestion
In this module, we will explore the pros and cons of using [smolagents](https://huggingface.co/docs/smolagents/en/index), helping you make an informed decision about whether it's the right framework for your needs.
```

Reviewer: As you prefer on this one but I was thinking maybe it's better to put the documentation link but feel free to refuse this suggestion.

## What is `smolagents`?

`smolagents` is a simple yet powerful framework for building AI agents. It provides LLMs with the _agency_ to interact with the real world, such as searching or generating images.

As we learned in unit 1, AI agents are programs that use LLMs to generate **'thoughts'** based on **'observations'** to perform **'actions'**. Let's explore how this is implemented in smolagents.

### Key Advantages of `smolagents`
- **Simplicity:** Minimal code complexity and abstractions
- **Flexible LLM Support:** Works with any LLM through integration with Hugging Face tools and external APIs
- **Code-First Approach:** First-class support for Code Agents that directly write their actions in code
- **HF Hub Integration:** Seamless integration with Hugging Face Hub, allowing the use of Gradio Spaces as tools

> ### When to use smolagents

Reviewer: ```suggestion
### When to use smolagents?
```

With these advantages in mind, when should we use smolagents over other frameworks? 

Smolagents is ideal when:

> - You need a lightweight and minimal solution

Reviewer: ```suggestion
- You need a **lightweight and minimal solution.**
```

> - You want to experiment quickly without complex configurations

Reviewer: ```suggestion
- You want to **experiment quickly** without complex configurations.
```

> - Your application logic is straightforward

Reviewer: ```suggestion
- Your **application logic is straightforward.**
```

### Code vs. JSON Actions

> Unlike other frameworks where agents write actions in JSON that require parsing, `smolagents` focuses on direct tool calls in code, simplifying the execution process. The following diagram illustrates this difference:

Reviewer: ```suggestion
Unlike other frameworks where agents write actions in JSON that require parsing, `smolagents` **focuses on direct tool calls in code**, simplifying the execution process. 

The following diagram illustrates this difference:
```

> ![Code vs. JSON actions](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/code_vs_json_actions.png)

Reviewer: ```suggestion
![Code vs. JSON actions](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/code_vs_json_actions.png)

If you don't remember the difference between Code vs JSON Actions, check [Unit 1, Section Actions](https://huggingface.co/learn/agents-course/unit1/actions#actions-enabling-the-agent-to-engage-with-its-environment)
```

### Agent Types in `smolagents`

> Agents in `smolagents` operate as multi-step agents. Each [`MultiStepAgent`](https://huggingface.co/docs/smolagents/main/en/reference/agents#smolagents.MultiStepAgent) performs:

Reviewer: ```suggestion
Agents in `smolagents` operate as **multi-step agents**. 

Each [`MultiStepAgent`](https://huggingface.co/docs/smolagents/main/en/reference/agents#smolagents.MultiStepAgent) performs:
```

- **[HfApiModel](https://huggingface.co/docs/smolagents/main/en/reference/models#smolagents.HfApiModel):** Wraps Hugging Face's [InferenceClient](https://huggingface.co/docs/huggingface_hub/main/en/guides/inference) to support the [Inference API](https://huggingface.co/docs/api-inference/index) and [Inference Providers](https://huggingface.co/blog/inference-providers).
- **[LiteLLMModel](https://huggingface.co/docs/smolagents/main/en/reference/models#smolagents.LiteLLMModel):** Leverages [LitLLM](https://www.litellm.ai/) for lightweight model interactions.
- **[OpenAIServerModel](https://huggingface.co/docs/smolagents/main/en/reference/models#smolagents.OpenAIServerModel):** Connects to models compatible with the OpenAI API server.
- **[AzureOpenAIServerModel](https://huggingface.co/docs/smolagents/main/en/reference/models#smolagents.AzureOpenAIServerModel):** Supports integration with any Azure OpenAI deployment.

This flexibility ensures that developers can choose the most suitable model integration for their specific use cases.

> Join us as we dive deep into smolagents in the upcoming sections!

Reviewer: ```suggestion
Now that we understood why and when to use smolagents, let's dive deeper into this powerful library!
```

Reviewer: You can change the conclusion, I have no strong opinions
